### PR TITLE
Fix isIPv4String() functionality

### LINKED ISSUE #65213: isIPv4String Function validating strings that are not proper IPv4 strings
The isIPv4String function is returning true for values that are not valid IPv4 strings.   This causes problems downstream when attempting to further process/store this value as IPv4. 

**How to reproduce**
```
select isIPv4String('1.1.1.1\0000')
select isIPv4String('1.1.1.1\0000s4')

┌─isIPv4String('1.1.1.1\0000')─┐
│                            1 │
└──────────────────────────────┘
```
Of course, trying to use this value in other IP functions fails
```
select isIPAddressInRange('1.1.1.1\0000', '1.1.1.1/8')

 isIPAddressInRange DB::Exception: Neither IPv4 nor IPv6 address: '1.1.1.1000'
```
**Expected behavior**
This should return false as it is not a valid IPv4String.

### PR DESCRIPTION
**Pull request by maynk27**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
update DB::parseIPv4whole to check if the whole string was parsed rather checking for null in the end. 
Fixes #65213 
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by maynk27** [comment]
Fixes #65213

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 1e80ffb19906e44f713679b9c7bcf9aae4e5dbcf with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65387/1e80ffb19906e44f713679b9c7bcf9aae4e5dbcf/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by superdiaodiao — COMMENTED** [review]
It would be better adding the SQL reported from the issue to the test cases. This will help confirm that the bug has been fixed.

**Comment by maynk27** [comment]
> It would be better adding the SQL reported from the issue to the test cases. This will help confirm that the bug has been fixed.

Could you please provide a link to the guide to add the test cases?

**Comment by superdiaodiao** [comment]
> > It would be better adding the SQL reported from the issue to the test cases. This will help confirm that the bug has been fixed.
> 
> Could you please provide a link to the guide to add the test cases?
> 

You can refer its feature PR and find the existing tests: https://github.com/ClickHouse/ClickHouse/pull/18349

**Comment by maynk27** [comment]
I need more time to figure this out. I suspect that the input after '\0' isn't being considered. For example https://ideone.com/lKlP39. I think the best approach would be to attach a debugger and see where exactly we are losing the data.

 I will keep this thread updated, meanwhile, suggestions are welcome. :)

**Comment by jsc0218** [comment]
You can read more from [here](https://clickhouse.com/docs/ru/development/tests), a test is needed for fixing a bug so can pass the ci.

**Comment by maynk27** [comment]
> You can read more from [here](https://clickhouse.com/docs/ru/development/tests), a test is needed for fixing a bug so can pass the ci.

I have added the test cases, the test is failing as the null character in the input string is not reaching the C++ layer.

File: src/Common/formatIPv6.h [hunk 1]
```diff
@@ -88,15 +88,16 @@ inline bool parseIPv4(T * &src, EOFfunction eof, unsigned char * dst, int32_t fi
...
+    if (parseIPv4(src, [&src, end](){ return src == end || *src == '\0'; }, dst))
         return src;
     return nullptr;
 }
 
 /// returns true if whole buffer was parsed successfully
 inline bool parseIPv4whole(const char * src, const char * end, unsigned char * dst)
 {
-    return parseIPv4(src, end, dst) == end;
+    const char * parsed = parseIPv4(src, end, dst);
```
Comment (alexey-milovidov) [inline src/Common/formatIPv6.h:99]: ?

File: src/Common/formatIPv6.h [hunk 2]
```diff
@@ -88,15 +88,16 @@ inline bool parseIPv4(T * &src, EOFfunction eof, unsigned char * dst, int32_t fi
 /// returns pointer to the right after parsed sequence or null on failed parsing
 inline const char * parseIPv4(const char * src, const char * end, unsigned char * dst)
 {
-    if (parseIPv4(src, [&src, end](){ return src == end; }, dst))
+    if (parseIPv4(src, [&src, end](){ return src == end || *src == '\0'; }, dst))
```
Comment (alexey-milovidov) [inline src/Common/formatIPv6.h:91]: Looks wrong.

**Comment by alexey-milovidov** [comment]
The fix is wrong. I will resubmit it.

**Pull request closed without merging**